Hermes/appscale/hermes/stats/subscribers/profile.py

```diff
@@ -47,30 +55,380 @@ def receive(self, nodes_stats_dict):
+    Renders header according to include_lists in advance and
+    creates base directory for processes stats profile log.
+    It also reads header of summary file (if it exists) to identify
+    order of columns.
+
+    Args:
+      include_lists: an instance of IncludeLists describing which fields
+        of processes stats should be written to CSV log.
+    """
+    self._header = (
+      ['utc_timestamp']
+      + converter.get_stats_header(process_stats.ProcessStats, include_lists)
+    )
+    self._include_lists = include_lists
+    self._directory = path.join(PROFILE_LOG_DIR, 'processes')
+    self._summary_file_name_template = 'summary-{resource}.csv'
+    try:
+      makedirs(self._directory)
+    except OSError as os_error:
+      if os_error.errno == errno.EEXIST and path.isdir(self._directory):
+        pass
+      else:
+        raise
+    self._summary_columns = self._get_summary_columns()
 
   def receive(self, processes_stats_dict):
-    # TODO
-    pass
+    """ Implements receive method of base class. Saves newly produced
+    cluster processes stats to a list of CSV files.
+    One detailed file for each process on every node and 3 summary files.
+
+    Args:
+      processes_stats_dict: a dict with node IP as key and list of
+        ProcessesStatsSnapshot as value.
+    """
+    services_summary = collections.defaultdict(self.ServiceProcessesSummary)
+
+    for node_ip, snapshots in processes_stats_dict.iteritems():
+
+      if snapshots:
+        # Add summary using only the latest snapshot from the node
+        for proc in snapshots[-1].processes_stats:
+          # Add this process stats to service summary
+          service_name = proc.unified_service_name
+          if proc.application_id:
+            service_name = '{}-{}'.format(service_name, proc.application_id)
+          summary = services_summary[service_name]
+          summary.cpu_time += proc.cpu.system + proc.cpu.user
+          summary.resident_mem += proc.memory.resident
+          summary.unique_mem += proc.memory.unique
+
+      # Write detailed process stats using all snapshots received from the node
+      for snapshot in snapshots:
+        for proc in snapshot.processes_stats:
+          # Write stats of the specific process to its CSV file
+          with self._prepare_file(node_ip, proc.monit_name) as csv_file:
+            writer = csv.writer(csv_file)
+            row = (
+              [snapshot.utc_timestamp]
+               + converter.stats_to_list(proc, self._include_lists)
+            )
+            writer.writerow(row)
+
+    # Update self._summary_columns ordered dict (set)
+    for service_name in services_summary:
+      if service_name not in self._summary_columns:
+        self._summary_columns.append(service_name)
+
+    # Write summary
+    self._save_summary(services_summary)
+
+  def _prepare_file(self, node_ip, monit_name):
+    """ Prepares CSV file with name processes/<node-IP>/<monit-name>.csv
```

Comment:
So what exactly is the file structure for the profiled stats if they're all running?